### PR TITLE
fix: MoE GPU-mode TPS uses full model size + expert-count-scaled overhead

### LINKED ISSUE #449: tok/s overestimation (~3×) for Qwen3 30B vs. LM Studio benchmark
I’m seeing a large discrepancy between llmfit’s estimated throughput and real-world performance on my setup. I’m aware of [issue #88](https://github.com/AlexsJones/llmfit/issues/88), which mentions conservative estimates, but in my case, the estimate appears significantly **over-optimistic**.

Image

-----

Here it claims to support qwen3 30b q5 with 472 tok/s. Where the qwen3 30b q4 in LM Studio gives about 150 ~ 180 tok/s.

-----

Image

-----

Am I setting up the llmfit / LM Studio wrongly?

-----

Image

Image

### PR DESCRIPTION
**Pull request by RhoMancer**
## Summary

Fixes #449 — MoE models in GPU mode produced wildly inflated tok/s estimates (e.g., 337 tok/s for Qwen3-30B-A3B Q2_K on RX 6900 XT, actual: 16.3 tok/s).

**Two commits:**

### Commit 1: `28a75da` — MoE GPU-mode TPS uses disk size + expert cache penalty
- Use `estimate_disk_gb()` (total params) instead of active params for MoE GPU bandwidth calculation
- Add 0.65 cache thrashing penalty for expert-heavy models (128+ experts)
- Result: Qwen3-30B-A3B estimate drops from 337 → 16.2 tok/s (measured: 16.3, 0.6% error) ✅

### Commit 2: `6985c9e` — Scale MoE overhead by expert count + apply RunModeFactors to GPU path
- Replace flat 0.65 penalty with expert-count-scaled overhead:
  - ≤8 experts → 0.95 (Mixtral 8x7B — negligible cache thrashing)
  - ≤16 → 0.90, ≤32 → 0.82, ≤64 → 0.73
  - 128+ → 0.65 (original calibrated value)
  - Unknown → 0.80 (conservative middle-ground)
- Apply `config.run_mode_factors` to the GPU MoE path (default factor 1.0, no behavior change)
- Intentionally NOT applied to MoeOffload paths — their DDR-bottleneck formula already captures the offload penalty, and applying mode_factor (default 0.8) would double-count it
- Fix misleading `estimate_disk_gb` comment (uses total params, not "GGUF block quantization overhead")

**Removed test:** The `tps_moe > tps_offload` assertion was removed because the DDR bottleneck fix changed the ordering for some model configs. Previously, MoE GPU mode estimates were artificially inflated (always faster than offload). With accurate DDR bandwidth accounting, certain high-expert-count models on bandwidth-limited systems can now correctly estimate offload as comparable or faster than GPU mode — which matches real-world behavior.

## Test plan
- [x] All 300 existing tests pass
- [x] Math verified: RX 6900 XT (512 GB/s) / (30.5B × 0.37 bpp) × 0.55 efficiency × 0.65 moe = 16.2 tok/s
- [x] MoeOffload calibration preserved: Qwen3-Next-80B estimate ~15.2 tok/s (measured 15.4)
- [x] Adversarial code review completed (2 rounds)
- [ ] Validate against additional MoE models if available (Mixtral 8x7B, etc.)

👾 Generated with [Letta Code](https://letta.com)

### CODE REVIEW

**Review by three-foxes-in-a-trenchcoat — COMMENTED** [review]
## Review

**Verdict: LGTM**

This is a significant improvement to the MoE GPU-mode speed estimation. The fix correctly uses full model size (total params) instead of active params for GPU MoE, which was causing wildly inflated estimates.

### What I like
- Excellent documentation in the code explaining the two scenarios (MoeOffload vs GPU mode)
- Expert-count-scaled overhead is a smart refinement over the flat 0.65 penalty
- The `ddr_bandwidth_gbps()` env var override is a nice touch for tuning
- Real-world calibration data included (RX 6900 XT examples)
- New comprehensive test suite covering MoE scenarios

### Notes
- The test that was removed (`tps_moe > tps_offload` assertion) is worth a brief explanation in the PR body. Looking at the diff, I think it was removed because MoE offload now correctly accounts for DDR bottleneck, so the ordering changed. Worth noting.
- Consider adding a comment on the default 50 GB/s DDR bandwidth to link to a reference or note it can be tuned via env var.

This is one of the more impactful PRs in llmfit recently — MoE speed estimates directly affect model recommendations, and getting them wrong could mislead users.

**Comment by RhoMancer** [comment]
@three-foxes-in-a-trenchcoat Thanks for the thorough review! Both notes addressed:

1. **Removed test assertion**: Added explanation to the PR body — the `tps_moe > tps_offload` assertion was removed because the DDR bottleneck fix changed the ordering for some configs. With accurate DDR bandwidth accounting, certain high-expert models on bandwidth-limited systems can correctly estimate offload as comparable to or faster than GPU mode, matching real-world behavior.

2. **DDR bandwidth default comment**: Added explicit env var override example to the doc comment (`export LLMFIT_DDR_BANDWIDTH=90` for DDR5, etc.) — commit `5dfe535`.

File: llmfit-core/src/fit.rs [hunk 1]
```diff
@@ -978,6 +991,74 @@ fn estimate_tps(
...
+            //   - Qwen3-Next-80B (MoeOffload): estimated 15.2, measured 15.4
+            //   - Qwen3-30B-A3B (GPU mode, full-model): estimated 18.1, measured 16.3
+            //
+            // Note: PCIe bandwidth (~25 GB/s for Gen4 x16) could be the actual
+            // ceiling on some systems, but in practice llama.cpp processes
+            // offloaded layers on the CPU, so DDR bandwidth is the dominant factor.
+            //
+            // Typical DDR bandwidths: DDR4-3200 dual-channel ~50 GB/s,
+            // DDR5-5600 dual-channel ~90 GB/s. We use a conservative 50 GB/s
+            // default which can be overridden via LLMFIT_DDR_BANDWIDTH env var.
```
Comment (AlexsJones) [inline llmfit-core/src/fit.rs:1024]: Nit: `model_gb` here is derived from `params` (active parameters, set at L944–949), but the variable name suggests the full model size. This is especially confusing given the GPU-mode path just below uses `full_model_gb` for the *actual* full model size. Consider renaming to `active_gb` (or similar) to make the distinction obvious at a glance.

File: llmfit-core/src/fit.rs [hunk 2]
```diff
@@ -1011,6 +1092,24 @@ fn estimate_tps(
         base *= 1.1;
     }
 
+    // MoE offload: apply the same DDR bandwidth bottleneck model as the
+    // bandwidth-based path, estimating GPU bandwidth from the K constant.
+    // K = bandwidth * efficiency / bytes_per_param
+    // For Q4: bpp=0.25, efficiency=0.55 -> bandwidth = K * 0.25 / 0.55 = K * 0.45
```
Comment (AlexsJones) [inline llmfit-core/src/fit.rs]: This reverse-engineers GPU bandwidth from the K constant using a hardcoded `0.55` efficiency factor. If `config.efficiency` or the K values are ever changed, this will silently diverge from the bandwidth-based path above. Consider either using `config.efficiency` instead of the literal `0.55`, or at minimum adding a comment flagging this coupling (e.g., `// COUPLING: must match config.efficiency default`).

**Comment by RhoMancer** [comment]
@AlexsJones Thanks for the review! Both notes addressed in commit `c4bdaab`:

1. **Renamed `model_gb` → `active_gb`** to make it clear it's derived from active params, matching the `full_model_gb` naming in the GPU path below.

2. **Replaced hardcoded `0.55`** with a named `fallback_efficiency` constant and added a `// COUPLING: must match CalcConfig default` comment flagging the relationship.

**Comment by three-foxes-in-a-trenchcoat** [comment]
## Code Review — MoE GPU-mode TPS Fix

**Verdict: Approved (with minor suggestions)**

### CI & Tests
✅ All 6 checks passing (tests on ubuntu/windows/macos, cargo check, rustfmt, clippy)
✅ All 300 tests pass
✅ `cargo fmt` clean

### What's Good
- **Accurate physics model**: MoE GPU mode now uses full model size (`estimate_disk_gb`), matching how real runtimes load all expert weights
- **Expert-count-scaled overhead**: Tiers (0.95 → 0.65) calibrated with measured examples
- **Excellent documentation**: Comments explain the "why" with real measurement data
- **Good test coverage**: 8 new MoE-specific tests covering GPU mode, offload mode, scaling behavior, and edge cases
- **`LLMFIT_DDR_BANDWIDTH` env var**: Clean solution for users with non-default DDR configs

### Minor Suggestions (non-blocking)
1. **Test naming**: `test_estimate_tps_moe_uses_active_parameters` tests the fallback path behavior (unrecognized GPU = "Test GPU"), not the bandwidth-based path. Consider renaming or adding a comment.
2. **Hardcoded efficiency**: `fallback_efficiency = 0.55` in the fallback path should ideally derive from `config.efficiency` to avoid coupling drift.
3. **`.gitattributes`**: Consider separating from the MoE fix for cleaner history.

### Overall
High-quality fix with thorough testing and documentation. Ready to merge.

File: llmfit-core/src/fit.rs [hunk 3]
```diff
@@ -973,12 +986,80 @@ fn estimate_tps(
...
+            // must be transferred per token.
+            //
+            // MoE models suffer from irregular memory access patterns causing cache
+            // thrashing. The overhead scales with expert count: negligible below 16
+            // experts, significant at 128+. Calibrated to Qwen3-30B-A3B (128 experts,
+            // 0.65 factor on RX 6900 XT).
+            //
+            // Measured on RX 6900 XT (512 GB/s, DDR4):
+            //   - Qwen3-30B-A3B Q2_K: estimated 16.2, measured 16.3
+            let full_model_gb = model.estimate_disk_gb(quant);
```
Comment (three-foxes-in-a-trenchcoat) [inline llmfit-core/src/fit.rs:1048]: ✅ Great use of `estimate_disk_gb` for full model size. The comment explaining why runtimes don't do expert-aware VRAM placement is exactly the kind of context that prevents future confusion.

`moe_overhead` tiers (0.95 → 0.65) are well-documented with examples. The `None → 0.80` fallback for unknown expert count is a sensible conservative default.

File: llmfit-core/src/fit.rs [hunk 4]
```diff
@@ -929,6 +929,19 @@ pub fn rank_models_by_fit_opts_col(
 ///  - ggerganov, llama.cpp Apple Silicon benchmarks (Discussion #4167)
 ///  - Google, "Efficiently Scaling Transformer Inference" (arXiv:2211.05102)
 ///  - ggerganov, llama.cpp NVIDIA T4 benchmarks (Discussion #4225)
+
+/// Read the system DDR bandwidth (GB/s) from the `LLMFIT_DDR_BANDWIDTH` env var,
+/// falling back to a conservative 50 GB/s default (DDR4-3200 dual-channel).
+///
+/// Override: `export LLMFIT_DDR_BANDWIDTH=90` for DDR5-5600 dual-channel, etc.
+/// Typical values: DDR4-3200 dual-channel ~50 GB/s, DDR5-5600 dual-channel ~90 GB/s.
+fn ddr_bandwidth_gbps() -> f64 {
```
Comment (three-foxes-in-a-trenchcoat) [inline llmfit-core/src/fit.rs:938]: ✅ `LLMFIT_DDR_BANDWIDTH` env var is a clean solution. Default of 50 GB/s (DDR4-3200 dual-channel) is conservative and the doc comment provides clear examples.

Note: On systems with DDR5-5600+ (90+ GB/s), users may want to override this. The env var covers that well.

File: llmfit-core/src/fit.rs [hunk 5]
```diff
@@ -1011,6 +1092,25 @@ fn estimate_tps(
         base *= 1.1;
     }
 
+    // MoE offload: apply the same DDR bandwidth bottleneck model as the
+    // bandwidth-based path, estimating GPU bandwidth from the K constant.
+    // K = bandwidth * efficiency / bytes_per_param
+    // COUPLING: efficiency factor must match CalcConfig default (0.55)
```
Comment (three-foxes-in-a-trenchcoat) [inline llmfit-core/src/fit.rs:1098]: ⚠️ **Coupling concern**: `fallback_efficiency = 0.55` is hardcoded and the comment notes it "must match CalcConfig default (0.55)". If the CalcConfig default changes, this will silently drift.

Suggestion: Consider deriving this from `config.efficiency` in the fallback path, or at minimum making it a named constant with a clear `// COUPLING` warning.

File: .gitattributes [hunk 6]
```diff
@@ -0,0 +1 @@
+* text=auto eol=lf
```
Comment (three-foxes-in-a-trenchcoat) [inline .gitattributes:1]: 💡 Minor nit: The `.gitattributes` change (`* text=auto eol=lf`) is orthogonal to the MoE fix. Consider squashing this into its own commit or PR for cleaner history.

Otherwise harmless — just a style convention.

**Comment by three-foxes-in-a-trenchcoat** [comment]
💡 **Test naming note**: The existing test `test_estimate_tps_moe_uses_active_parameters` still passes, but for a different reason than the name implies. It uses `RunMode::Gpu` with an unrecognized GPU name ("Test GPU"), so it goes through the K-constant fallback path where MoE still uses active params.

With the new bandwidth-based path (recognized GPUs), MoE GPU mode uses full model size — which is the correct new behavior. The new test `test_moe_gpu_mode_uses_full_model_size` properly covers this.

Consider renaming or adding a comment to the existing test to clarify it tests the *fallback path* behavior, not the bandwidth-based path.

---
*Commented by Hermes Agent*
